### PR TITLE
Fix 'Use Local Function' fixer in code style layer

### LINKED ISSUE #68440: CodeFixProvider throws InvalidOperationException: 'ISymbolDeclarationService' is required to accomplish the task but is not available from the workspace
**Version Used**: 
VS 17.6.2

**Steps to Reproduce**:

1. Create a new .NET 7 console app.
2. Add the line `Func<int, int> test = (int n) => n * 2;`.
3. Place the cursor at the new line and open the quick actions menu. Note that the quick action "Use local function" appears and shows a correct preview.
4. Add the line `<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>` to the project file.
5. Open the quick actions menu on the code line again.

**Diagnostic Id**:
IDE0039

**Expected Behavior**:
The code fix "Use local function" should be available.

**Actual Behavior**:
VS shows a yellow error bar with the error "'CSharpUseLocalFunctionCodeFixProvider' encountered an error and has been disabled." and the following stack trace:

```
System.InvalidOperationException : Service of type 'Microsoft.CodeAnalysis.LanguageService.ISymbolDeclarationService' is required to accomplish the task but is not available from the workspace.
   at Microsoft.CodeAnalysis.Host.HostLanguageServices.GetRequiredService[TLanguageService]()
   at async Microsoft.CodeAnalysis.CSharp.UseLocalFunction.CSharpUseLocalFunctionCodeFixProvider.FixAllAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeFixes.SyntaxEditorBasedCodeFixProvider.FixAllWithEditorAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetChangedSolutionAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputeOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.PreviewChangesSuggestedAction.CreateAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.GetPreviewChangesFlavorAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.CreateAllFlavorsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #68440

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs [hunk 1]
```diff
@@ -93,7 +93,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             {
 #if CODE_STYLE
                 var info = new CSharpCodeGenerationContextInfo(
-                    CodeGenerationContext.Default, CSharpCodeGenerationOptions.Default, new CSharpCodeGenerationService(document.Project.Services), root.SyntaxTree.Options.LanguageVersion());
+                    CodeGenerationContext.Default, CSharpCodeGenerationOptions.Default, new CSharpCodeGenerationService(document.Project.GetExtendedLanguageServices().LanguageServices), root.SyntaxTree.Options.LanguageVersion());
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs:96]: 📝 This fixes the root cause of the bug

File: src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs [hunk 2]
```diff
@@ -24,7 +24,7 @@ public UseLocalFunctionTests(ITestOutputHelper logger)
         }
 
         internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderAndFixer(Workspace workspace)
-            => (new CSharpUseLocalFunctionDiagnosticAnalyzer(), GetCSharpUseLocalFunctionCodeFixProvider());
+            => (new CSharpUseLocalFunctionDiagnosticAnalyzer(), new CSharpUseLocalFunctionCodeFixProvider());
```
Comment (sharwell) [inline src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs:27]: 📝 This fixes the failure to identify the bug in testing

File: src/Analyzers/CSharp/Tests/MakeLocalFunctionStatic/MakeLocalFunctionStaticTests.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ public MakeLocalFunctionStaticTests(ITestOutputHelper logger)
         }
 
         internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderAndFixer(Workspace workspace)
-            => (new MakeLocalFunctionStaticDiagnosticAnalyzer(), GetMakeLocalFunctionStaticCodeFixProvider());
+            => (new MakeLocalFunctionStaticDiagnosticAnalyzer(), new MakeLocalFunctionStaticCodeFixProvider());
```
Comment (sharwell) [inline src/Analyzers/CSharp/Tests/MakeLocalFunctionStatic/MakeLocalFunctionStaticTests.cs:28]: 📝 This is another code fixer which was not properly tested, but there were no test failures when it was corrected

File: src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs [hunk 4]
```diff
@@ -93,7 +93,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
 #else
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs:97]: so, this seems like something we likely need an analyzer for, right?  Like it's *farrrrrr* too easy to screw this up.  

Another option seems to be `(CSharpCodeGenarationService)document.GetRequiredLanguageService<ICodeGenerationService>()` (or just make CSharpCodeGenerationContextInfo take an ICodeGenService, i dont' think it needs the subtype.

File: src/EditorFeatures/DiagnosticsTestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs [hunk 5]
```diff
@@ -249,12 +249,6 @@ private static void AssertNoAnalyzerExceptionDiagnostics(IEnumerable<Diagnostic>
         // This enables porting the tests for the ported analyzer in CodeStyle layer.
         #region CodeFixProvider Helpers
 
-        // https://github.com/dotnet/roslyn/issues/43056 blocks porting the fixer to CodeStyle layer.
-        protected static CodeFixProvider GetMakeLocalFunctionStaticCodeFixProvider() => new MakeLocalFunctionStaticCodeFixProvider();
-
-        // https://github.com/dotnet/roslyn/issues/43056 blocks porting the fixer to CodeStyle layer.
-        protected static CodeFixProvider GetCSharpUseLocalFunctionCodeFixProvider() => new CSharpUseLocalFunctionCodeFixProvider();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/DiagnosticsTestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs:256]: w00t.

File: src/EditorFeatures/DiagnosticsTestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs [hunk 6]
```diff
@@ -249,12 +249,6 @@ private static void AssertNoAnalyzerExceptionDiagnostics(IEnumerable<Diagnostic>
...
-
         // https://github.com/dotnet/roslyn/issues/43091 blocks porting the fixer to CodeStyle layer.
         protected static CodeFixProvider GetCSharpUseAutoPropertyCodeFixProvider() => new CSharpUseAutoPropertyCodeFixProvider();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/DiagnosticsTestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs:253]: ick.  i'll see if i could get these guys to move down too.

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm.  i think we could make this slightly nicer though by avoidign the direct casll to GetExtendedLanguageServices here.

File: src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs (hunk 4 above)
Comment (sharwell) [reply src/Analyzers/CSharp/CodeFixes/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs:97]: ➡️ Updated BannedSymbols.txt to account for the new `Services` property, which was not present at the time `LanguageServices` property was excluded.

File: src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs:27]: @sharwell I'm very surprised. The deleted `GetCSharpUseLocalFunctionCodeFixProvider` method exactly does `new CSharpUseLocalFunctionCodeFixProvider()`, how that would matter in identifying the bug in testing?

Comment (sharwell) [reply src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs:27]: There are two instances of this type. One in the code style layer, and one in the features layer. Previously, when tests compiled in the code style layer they were still using the fixer from the features layer. The relocated instantiation binds to the other type, so tests in the code style layer use the fix from the code style layer.

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/UseLocalFunction/UseLocalFunctionTests.cs:27]: Makes sense! Thanks @sharwell

**Pull request merged**